word_finder.cpp

```diff
@@ -0,0 +1,208 @@
+    for(unsigned int i = 0U ; i < size ; ++i) {
+        wchar_t charStart = word.at(i);
+        if ( L' ' == charStart ) {
+            continue;
+        }
+
+        auto nextNode = _root;
+        unsigned int end = numeric_limits<unsigned int>::max();
+
+        for(unsigned int j = i ; j < size ; ++j) {
+            wchar_t c = word.at(j);
+            if (c == L' ') {
+                continue;
+            }
+            nextNode = nextNode->nextNode(c);
+            if (!nextNode) {
+                break;
+            }
+
+            if (nextNode->isEndNode()) {
+                end = j;
+            }
+        }
+
+        if (numeric_limits<unsigned int>::max() != end) {
+            list.emplace_back(i, end);
+        }
+    }
+    return list;
+}
+
+wstring StringUtils::replace(wstring str, const wstring &from, const wstring &to) {
+    wstring::size_type begin = 0UL;
+    while((begin = str.find(from, begin)) != wstring::npos) {
+        str.replace(begin, from.length(), to);
+        begin += to.length();
+    }
+    return str;
+}
+
+wstring StringUtils::trim(const wstring &str) {
+    const wstring blank = L"\r\n\t ";
+    size_t begin = str.find_first_not_of(blank);
+    size_t end = str.find_last_not_of(blank);
+    if (wstring::npos != begin || wstring::npos != end) {
+        return str.substr(begin, end - begin + 1);
+    }else{
+        return L"";
+    }
+}
+
+wstring StringUtils::toLowerCase(wstring str) {
+    transform(str.begin(), str.end(), str.begin(), ::tolower);
+    return str;
+}
+
+const wchar_t * StringUtils::charsToWstring(const char* str) {
+    string currentLocale = setlocale(LC_ALL, "C.UTF-8");
+
+    size_t size = mbstowcs(nullptr, str, 0UL) + 1UL;
+    wchar_t *wstr = new wchar_t[size];
+    wmemset(wstr, 0, size);
+    mbstowcs(wstr, str, size);
+
+    setlocale(LC_ALL, currentLocale.c_str());
+    return wstr;
+}
+
+void StringUtils::free_charsToWstring(const wchar_t *str) {
+    if (nullptr != str) {
+        delete [] str;
+    }
+}
+
+const char * StringUtils::wstringToChars(const wstring &wstr) {
+    string currentLocale = setlocale(LC_ALL, "C.UTF-8");
+
+    const wchar_t *wchars = wstr.c_str();
+    size_t size = wcstombs(nullptr, wchars, 0UL) + 1UL;
+    char *str = new char[size];
+    memset(str, 0, size);
+    wcstombs(str, wchars, size);
+
+    setlocale(LC_ALL, currentLocale.c_str());
+    return str;
+}
+
+void StringUtils::free_stringToChars(const char *str) {
+    if (nullptr != str) {
+        delete [] str;
+    }
+}
+
+IllegalWordChecker::IllegalWordChecker()
+: _finder() {
+}
+
+IllegalWordChecker::IllegalWordChecker(const list<wstring> &strlist)
+: _finder(strlist) {
+}
+
+void IllegalWordChecker::add(const wstring &str) {
+    _finder.add(str);
+}
+
+wstring & IllegalWordChecker::filter(wstring &str) const {
+    list<Word> list;
+    for(const Word &word : _finder.find(list, str)) {
+        str.replace(word.start, word.count(), word.count(), L'*');
```

Comment:
The `L'*'` should be passed as parameter instead of hard-coded.
To be more general, the parameter can be string instead of character.